Trustroots/trustroots
# Migrate unread message count to react

#### Proposed Changes

* migrate the unread messages counter service and component to react

Originally it was just about the unread messages counter component, but I decided to also migrate the polling service away from angular too.

It's now a bit more separated, but potentially feels more complex as there are 3 seperate services. But they hopefully have clear roles! I'm a bit confused lately generally, so would be glad to hear from others if they find it makes sense or not, I don't know if I can tell any more :/

I made them all so they have enable/disable functions so we have more control over when they begin their operation (as opposed to starting as soon as the module is imported). At the moment the angular `run` block starts them.

The testing code was a bit annoying actually, as all the angular-compat stuff needs mocking away, and I needed `jest.isolateModules()` to probably reset the state of a service file between tests. The tests are somewhat minimal.

Also, regarding testing, to make them work with all the existing tests we basically bootstrap the core of the app automatically in `jest/jest.setup.js` but eventually it would be nice to to remove that setup and have each test loads what it needs.

The services are:

**unread mesage count service**

The job of this is to poll for unread messages counts when visible and logged in.

**visibility service**

Use the [Page Visiblity API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) to check if the page is in the foreground/background for two purposes:
- lower the polling rate when not visible
- refresh the count immediately when returning to foreground

The old way used the page blur events, this replaces that.

**favicon updater service**

Just updates the favicon when the count changes.

Also, involved adding some utilities:

**dom ready**

Calls the callback when the DOM is ready! (like jQuery function). Useful if doing DOM stuff - used here for the favicon updater...

```js
import { ready } from '@/modules/core/client/utils/dom';

ready(() => {
  const el = document.getElementById('foo');
});
```

**createSubscribable**

I like the simple pattern for have subscribe/notify behaviour.

```js
import createSubscribable from '@/modules/core/client/utils/subscribable';

 const { subscribe, notify } = createSubscribable();
 
 // add a subscriber, keeping hold of the unsubscribe function
 const unsubscribe = subscribe(value => console.log('received value', value));
 
 // notify all subscribers with a value
 notify('some value');

 // unsubscribe our subscriber
 unsubscribe();
```

There is probably some library in npm that does this, but they often have loads of extra features, etc...

#### Testing Instructions

* `git switch migrate/unread-message-count`
* `npm start`
* you need two accounts in seperate browsers/sessions
* write messages to each other, leaving them unread for one of the users
* see if the little red indicator on messages changes
  ![image](https://user-images.githubusercontent.com/7449720/79145851-a0e8e880-7dc1-11ea-8115-26cab4b722bc.png)
* see if the favicon changes
  ![image](https://user-images.githubusercontent.com/7449720/79146013-ee655580-7dc1-11ea-9e4b-e81a053efc5e.png)

The polling interval is 30s for foreground, and 300s for background, so be a little patient.

Fixes #
